utils/include/utils/merge_sorted.h

```diff
@@ -0,0 +1,160 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+#ifndef HLASMPLUGIN_UTILS_MERGE_SORTED_H
+#define HLASMPLUGIN_UTILS_MERGE_SORTED_H
+
+#include <algorithm>
+#include <compare>
+#include <concepts>
+#include <functional>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hlasm_plugin::utils {
+
+struct trivial_merger
+{
+    void operator()(auto&&, const auto&) const noexcept {}
+};
+
+template<typename T, typename TargetVector, typename SourceIterator>
+concept merge_key_comparator = std::is_same_v<std::strong_ordering,
+    std::invoke_result_t<T, typename TargetVector::reference, typename TargetVector::reference>> && std::
+    is_same_v<std::strong_ordering,
+        std::invoke_result_t<T,
+            typename TargetVector::reference,
+            typename std::iterator_traits<SourceIterator>::reference>>;
+
+template<typename T, typename TargetVector, typename SourceIterator>
+concept merge_updater =
+    std::invocable<T, typename TargetVector::reference, typename std::iterator_traits<SourceIterator>::reference>;
+
+template<typename T,
+    /* std::forward_iterator - move_iterator is viewed as input_iterator only before libc++17 (and possibly c++23) */
+    typename It,
+    std::sentinel_for<It> S,
+    merge_key_comparator<std::vector<T>, It> KeyComparator = std::compare_three_way,
+    merge_updater<std::vector<T>, It> Merger = trivial_merger>
+void merge_sorted(
+    std::vector<T>& sorted_vec, It it, const S ite, KeyComparator&& cmp = KeyComparator(), Merger&& m = Merger())
+{
+    constexpr const auto trivial_inserter = !std::invocable<Merger, typename std::iterator_traits<It>::reference>;
+    auto&& p = [&m]() -> std::conditional_t<trivial_inserter, std::identity, Merger&> {
+        if constexpr (trivial_inserter)
+            return std::identity();
+        else
+            return m;
+    }();
+    const auto init_size = sorted_vec.size();
+    for (size_t i = 0; i < init_size && it != ite;)
+    {
+        auto&& el = sorted_vec[i];
+        if (auto c = std::invoke(cmp, el, *it); c == 0)
+        {
+            std::invoke(m, el, *it);
+            ++i;
+            ++it;
+        }
+        else if (c > 0)
+            sorted_vec.push_back(std::invoke(p, *it++));
+        else
+            ++i;
+    }
+    // libc++: std::views::transform(std::views::subrange(it, ite), std::ref(p))...
+    if constexpr (trivial_inserter)
+        sorted_vec.insert(sorted_vec.end(), it, ite);
+    else
+    {
+        for (; it != ite; ++it)
+            sorted_vec.push_back(std::invoke(p, *it));
+    }
+
+    const auto less = [&cmp](const auto& l, const auto& r) { return std::invoke(cmp, l, r) < 0; };
+    std::inplace_merge(sorted_vec.begin(), sorted_vec.begin() + init_size, sorted_vec.end(), less);
```

Comment:
Can it happen that the two ranges (begin() -> middle and middle -> end()) are in fact not sorted which would break the condition to use `std::inplace_merge(...)`? 

I'm imagining a case when `std::invoke(m, el, *it);` is called few lines above which could alter the key against which we are comparing during `std::invoke(cmp, l, r)`. 

E.g. before calling `std::inplace_merge(...)`, the `sorted_vec` could look like this:
0,1,2,5,4,6,7,8,9 and `middle` would be `6`

Reply:
The first range (begin->middle) is always sorted - that's a precondition.

Other preconditions (not documented anywhere) should be that
- the update operation preserves order (just like map elements have const keys)
- the insert action/projection preservers order `cmp(el, *it) == cmp(el, m(*it))`

Reply:
I'll think about it.

Reply:
I thought about it and I think that the best I can do is to verify that the supplied functions are not "too evil", simply because the algorithm should work with move-only types and therefore the the values before the operation may not exist after the operation is finished.